File: 4/4.clj
(defn is-palindrome?
  [s]
  (loop [i 0]
    (def len (count s))
    (cond
      (>= i (/ len 2)) true
      (= (nth s i) (nth s (- len i 1))) (recur (inc i))
      :else false)))

(def three-digits (range 999 99 -1))
(apply max (for [x three-digits y three-digits
                 :let [product (* x y)]
                 :when (is-palindrome? (str product))]
             product))

File: 4/4.go
package main

import (
	"fmt"
	"strconv"
)

func isPalindrome(s string) bool {
	i, j := 0, len(s) - 1
	for i < j {
		if s[i] != s[j] {
			return false
		}
		i++
		j--
	}
	return true
}

func main() {
	max := 0
	for i := 100; i < 999; i++ {
		for j := 100; j < 999; j++ {
			product := i * j
			if isPalindrome(strconv.Itoa(product)) {
				if product > max {
					max = product
				}
			}
		}
	}
	fmt.Println(max)
}
